File: binary/python3/binary_tree_level_order_traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:        
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        queue = []
        traversalResult = []
        
        if root is None:
            return traversalResult
        
        queue.append(root)
        
        level = []
        levelSize = len(queue)
        traversedSize = 0
        
        for node in queue:
            level.append(node.val)
            if node.left is not None:
                queue.append(node.left)
                
            if node.right is not None:
                queue.append(node.right)
                
            traversedSize += 1
            
            if len(level) == levelSize:
                traversalResult.append(level)
                level = []
                levelSize = len(queue) - traversedSize

        return traversalResult

File: binary/cpp/binary_tree_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> answer;
        vector<int> levelAnswer;
        TreeNode* target;
        std::queue<TreeNode*> que;
        int levelSize;

        if (!root) {
            return answer;
        }
        que.push(root);
        while (!que.empty()) {
            levelSize = que.size();
            levelAnswer.clear();
            for (int i = 0; i < levelSize; i++) {
                target = que.front();
                levelAnswer.push_back(target->val);
                if (target->left) {
                    que.push(target->left);
                }
                if (target->right) {
                    que.push(target->right);
                }
                que.pop();
            }
            answer.push_back(levelAnswer);
        }
        return answer;
    }
};
